### PR TITLE
Helpers

### PR DESCRIPTION
**Pull request by aliang8**
Adding helper functions.

### CODE REVIEW

File: lib/query.js [hunk 1]
```diff
@@ -2,14 +2,18 @@
...
 export const isIdentifier = node => node && node.type === 'Identifier'
 export const isApply = node => node && node.type === 'Apply'
+export const isParens = node => node && node.type === 'Parentheses'
 
 export const isOperation = node => isApply(node) && !isNumber(node)
 export const isFunction = node => isApply(node) && isIdentifier(node.op)
 
 // TODO: curry it?
-const _isOp = (op, node) => isApply(node) && node.op === op
+export const _isOp = (op, node) => isApply(node) && node.op === op
```
Comment (k4b7) [inline lib/query.js:17]: Usually names prefixed with `_` are intended to be private.  If we want to make it publish we should change the name.  The current `isOperation` is functionally the same as `isApply` so I think we should make a breaking change and deprecate `isOperation` in its current form and replace `_isOp` with `isOperation`.  We could also call it `isOp` if `isOperation` is too long.

File: lib/query.js [hunk 2]
```diff
@@ -44,6 +48,47 @@ export const isNumber = node => {
     }
 }
 
+export const isInteger = node => {
+    return isNumber(node) && Number.isInteger(Number(node.value))
+}
+
+export const isConstantFraction = node => isFraction(node) && node.args.every(arg => isNumber(arg))
+export const isIntegerFraction = node => isFraction(node) && node.args.every(arg => isInteger(arg))
```
Comment (k4b7) [inline lib/query.js]: This can be simplified by doing `isFraction(node) && node.args.every(isInteger)`.  A similar change change be made to `isConstantFraction`.

File: lib/query.js [hunk 3]
```diff
@@ -44,6 +48,47 @@ export const isNumber = node => {
...
+
+export const isPolynomial = (node) => {
+    return isAdd(node) && node.args.every(isPolynomialTerm)
+}
+
+export const isVariableFactor = (node) =>
+    isIdentifier(node) ||
+    isPow(node) && isIdentifier(node.args[0]) &&
+    (isNumber(node.args[1]) || isVariableFactor(node.args[1]))
```
Comment (k4b7) [inline lib/query.js]: Can you indent this line so that it's clearer that this depends on the the node being a power?

File: lib/query.js [hunk 4]
```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
 
 export const getNumerator = (node) => isNeg(node) ? getNumerator(node.args[0]) : node.args[0]
 export const getDenominator = (node) => isNeg(node) ? getDenominator(node.args[0]) : node.args[1]
+
+// TODO: handle multivariable polynomials
+// Get degree of a polynomial term
+// e.g. 6x^2 -> 2
+export const getExponent = (node) => {
```
Comment (k4b7) [inline lib/query.js]: Can you change the name of this function to `getPolyDegree`?

File: lib/query.js [hunk 5]
```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
...
+        return 0
+    } else if (isIdentifier(node) || isPolynomial(node)){
+        return 1
+    } else if (isPow(node)) {
+        return getValue(node.args[1])
+    } else if (isMul(node)){
+        return getExponent(node.args[1])
+    } else if (isNeg(node)) {
+        const variable = node.args[0]
+        return getExponent(variable.args[1])
```
Comment (k4b7) [inline lib/query.js]: Should this be `return getExponent(nodes.args[0])`?

File: lib/query.js [hunk 6]
```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
...
+    }
+}
+
+export const getCoefficient = (node) => {
+    if (isNumber(node)) {
+        return node
+    } else if (isIdentifier(node) || isPow(node)) {
+        return build.numberNode(1)
+    } else if (isNeg(node)) {
+        const result = build.applyNode('neg', [getCoefficient(node.args[0])])
```
Comment (k4b7) [inline lib/query.js]: It's unfortunate that `applyNode` has such a different API.  I'll do some thinking on how to make it better.

Comment (k4b7) [reply lib/query.js]: I'd expect calling this on something like `-2x^2` to return `-2` but I think as it's written it only returns `2`.

Comment (k4b7) [reply lib/query.js]: We should probably have tests for some of these larger functions.  I can set up test infra for what's already on `master`.

File: lib/query.js [hunk 7]
```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
...
+        result.wasMinus = node.wasMinus
+        return result
+    } else if (isMul(node)) {
+        const numbers = node.args.filter(arg => isNumber(arg) || isConstantFraction(arg))
+        if (numbers.length > 1) {
+            return build.applyNode('mul', numbers)
```
Comment (k4b7) [inline lib/query.js]: `build.mul(...numbers)`

File: lib/query.js [hunk 8]
```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
...
+        } else if (numbers.length > 0) {
+            return numbers[0]
+        } else {
+            return build.numberNode(1)
```
Comment (k4b7) [inline lib/query.js]: `build.number('1')`... I need to update `buildNumber`/`number` to convert numbers to strings.

File: lib/query.js [hunk 9]
```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
...
+export const getVariableFactors = (node) => {
+    if (isVariableFactor(node)) {
+        return [node]
+    } else if (isMul(node)) {
+        return node.args.filter(isVariableFactor)
+    } else if (isNeg(node)) {
+        // TODO: figure out how to handle (x)(-y)(z)
+        return getVariableFactors(node.args[0])
+    } else {
+        // throw?
```
Comment (k4b7) [inline lib/query.js]: I think it's fine to return an empty array here.

File: lib/query.js [hunk 10]
```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
...
+}
+
+export const getVariableFactors = (node) => {
+    if (isVariableFactor(node)) {
+        return [node]
+    } else if (isMul(node)) {
+        return node.args.filter(isVariableFactor)
+    } else if (isNeg(node)) {
+        // TODO: figure out how to handle (x)(-y)(z)
+        return getVariableFactors(node.args[0])
```
Comment (k4b7) [inline lib/query.js:165]: Maybe `isVariable` factor should also return true for `-x`.

File: lib/query.js [hunk 11]
```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
...
+    }
+}
+
+export const getVariableFactorName = (node) => {
+    if (isIdentifier(node)) {
+        return node.name
+    } else if (isPow(node)) {
+        return getVariableFactorName(node.args[0])
+    } else {
+        // throw?
```
Comment (k4b7) [inline lib/query.js]: I think returning `null` would be sufficient.

Comment (k4b7) [reply lib/query.js]: We can make things safer by introducing static typing.

File: lib/query.js [hunk 12]
```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
...
+    }
+}
+
+export const sortVariables = (variables) =>
+    variables.sort(
+        (a, b) => getVariableFactorName(a) > getVariableFactorName(b))
+
+// TODO: fix the case of single polynomials ex. 2x^2
```
Comment (k4b7) [inline lib/query.js]: This should be pretty easy to fix.  Can you fix it as part of this PR or put it in a separate PR?

File: lib/query.js [hunk 13]
```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
...
+    })
+
+    return {coefficientMap, constants}
+}
+
+// returns true if two nodes have the same base variable
+// e.g. (x + 1)^1 and (x + 1)^3 , true
+// e.g. x^2, (x + 1)^2, false
+export const hasSameBase = (node1, node2) => {
+    return print(node1) == print(node2)
```
Comment (k4b7) [inline lib/query.js:218]: This is insufficient as `x + y !== y + x`

Comment (k4b7) [reply lib/query.js:218]: Really we should have an `nodeEquals` function that traverse both nodes.  We have something that almost does this in `math-rules`.  I can add a `nodeEquals` function.

File: lib/query.js (hunk 1 above)
Comment (aliang8) [reply lib/query.js:17]: Oh this was an accident. We could keep `_isOp` as it is since it seems to be working fine.

File: lib/query.js (hunk 2 above)
Comment (aliang8) [reply lib/query.js]: Gotcha!

File: lib/query.js (hunk 5 above)
Comment (aliang8) [reply lib/query.js]: I think for this, I was under the assumption that the original node is of the form `-2x^2` (it has a coefficient).

File: lib/query.js (hunk 6 above)
Comment (aliang8) [reply lib/query.js]: Great, that would be very helpful!

File: lib/query.js (hunk 12 above)
Comment (aliang8) [reply lib/query.js]: Yup, I had it fixed before but I think I lost my changes :(

File: lib/query.js (hunk 13 above)
Comment (aliang8) [reply lib/query.js:218]: Great!
